File: benchmarks/python/set_ops.py
"""Set Operations Benchmark for Python"""

def bench_build(n, iters):
    s = set()
    for _ in range(iters):
        s = set()
        for i in range(n):
            s.add(i)
    return s

def bench_contains_hit(s, n, iters):
    hits = 0
    for _ in range(iters):
        for i in range(n):
            if i in s:
                hits += 1
    return hits

def bench_contains_miss(s, n, iters):
    misses = 0
    for _ in range(iters):
        for i in range(n, n + n):
            if i not in s:
                misses += 1
    return misses

def bench_union(a, b, iters):
    checksum = 0
    for _ in range(iters):
        result = a | b
        checksum += len(result)
    return checksum

def bench_intersect(a, b, iters):
    checksum = 0
    for _ in range(iters):
        result = a & b
        checksum += len(result)
    return checksum

def bench_difference(a, b, iters):
    checksum = 0
    for _ in range(iters):
        result = a - b
        checksum += len(result)
    return checksum

if __name__ == "__main__":
    N = 10000
    s = bench_build(N, 200)
    print(f"build: done")
    hits = bench_contains_hit(s, N, 200)
    print(f"contains_hit: {hits}")
    misses = bench_contains_miss(s, N, 200)
    print(f"contains_miss: {misses}")
    a = set(range(5000))
    b = set(range(5000, 10000))
    union_count = bench_union(a, b, 500)
    print(f"union: {union_count}")
    c = set(range(N))
    half = N // 2
    d = set(range(half, half + N))
    intersect_count = bench_intersect(c, d, 500)
    print(f"intersect: {intersect_count}")
    diff_count = bench_difference(c, d, 500)
    print(f"difference: {diff_count}")

File: benchmarks/go/set_ops.go
package main

import "fmt"

func benchBuild(n int, iters int) map[int]struct{} {
	var s map[int]struct{}
	for iter := 0; iter < iters; iter++ {
		s = make(map[int]struct{})
		for i := 0; i < n; i++ {
			s[i] = struct{}{}
		}
	}
	return s
}

func benchContainsHit(s map[int]struct{}, n int, iters int) int {
	hits := 0
	for iter := 0; iter < iters; iter++ {
		for i := 0; i < n; i++ {
			if _, ok := s[i]; ok {
				hits++
			}
		}
	}
	return hits
}

func benchContainsMiss(s map[int]struct{}, n int, iters int) int {
	misses := 0
	for iter := 0; iter < iters; iter++ {
		for i := n; i < n+n; i++ {
			if _, ok := s[i]; !ok {
				misses++
			}
		}
	}
	return misses
}

func benchUnion(a map[int]struct{}, b map[int]struct{}, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		result := make(map[int]struct{})
		for k := range a {
			result[k] = struct{}{}
		}
		for k := range b {
			result[k] = struct{}{}
		}
		checksum += len(result)
	}
	return checksum
}

func benchIntersect(a map[int]struct{}, b map[int]struct{}, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		result := make(map[int]struct{})
		for k := range a {
			if _, ok := b[k]; ok {
				result[k] = struct{}{}
			}
		}
		checksum += len(result)
	}
	return checksum
}

func benchDifference(a map[int]struct{}, b map[int]struct{}, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		result := make(map[int]struct{})
		for k := range a {
			if _, ok := b[k]; !ok {
				result[k] = struct{}{}
			}
		}
		checksum += len(result)
	}
	return checksum
}

func main() {
	const N = 10000

	s := benchBuild(N, 200)
	fmt.Println("build: done")

	hits := benchContainsHit(s, N, 200)
	fmt.Printf("contains_hit: %d\n", hits)

	misses := benchContainsMiss(s, N, 200)
	fmt.Printf("contains_miss: %d\n", misses)

	// Build two 5k-element sets for union
	a := make(map[int]struct{})
	b := make(map[int]struct{})
	for i := 0; i < 5000; i++ {
		a[i] = struct{}{}
	}
	for i := 5000; i < 10000; i++ {
		b[i] = struct{}{}
	}

	unionCount := benchUnion(a, b, 500)
	fmt.Printf("union: %d\n", unionCount)

	// Build two overlapping 10k-element sets for intersect/difference
	c := make(map[int]struct{})
	d := make(map[int]struct{})
	for i := 0; i < N; i++ {
		c[i] = struct{}{}
	}
	half := N / 2
	for i := half; i < half+N; i++ {
		d[i] = struct{}{}
	}

	intersectCount := benchIntersect(c, d, 500)
	fmt.Printf("intersect: %d\n", intersectCount)

	diffCount := benchDifference(c, d, 500)
	fmt.Printf("difference: %d\n", diffCount)
}
